Add `nonzero_static()` op to pytorch to unblock export

guangy10 opened the pull request:
Summary: Add new experimental python op (`torch.nonzero_static`) for export. There is NO cuda impl included in this PR

Example:

Say input tensor is `x = torch.tensor([[1, 0], [3, 2]])`

call regular `nonzero()` on x will give you a tensor `tensor([[0, 0], [1, 0], [1, 1])`
call `nonzero_static(x, size=4)` on x will give you a tensor `tensor([[0, 0], [1, 0], [1, 1], [fill_value, fill_value])` (padded)
call `nonzero_static(x, size=2)` on x will give you a tensor `tensor([[0, 0], [1, 0])` (truncated)

Test Plan:
**Unit Tests**
```
buck test @mode/dev-nosan //caffe2/test:test_dynamo -- 'caffe2/test:test_dynamo - test_export.py::ExportTests::test_export_with_nonzero_static' -- 'caffe2/test:test_dynamo - test_misc.py::MiscTests::test_nonzero_static'
```

**PT2 Export with `nonzero_static()`**
Example of `GraphModule` in the exported graph
```
def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    nonzero_static_default = torch.ops.aten.nonzero_static.default(arg0, size = 4);  arg0 = None
    return pytree.tree_unflatten([nonzero_static_default], self._out_spec)
```

Differential Revision: D44324808



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2234,6 +2234,33 @@ Tensor nonzero_cpu(const Tensor& self) {
+    return at::empty(0, ScalarType::Long);
+  }
+  // Delegate call to regular nonzero to get a data-dependent output
+  auto dyn_result = at::nonzero(self);
```
No. If you're doing this in C++, you must implement this properly (with no syncs). The whole reason I told you to register it via Python dispatcher was so that the crappy implementation wasn't available by default to users. People will hit this version, and will reasonable start asking questions about why it is syncing even with an explicit size.

ezyang reviewed (dismissed):
PR as written introduces performance footgun.

ezyang commented on functorch/experimental/_nonzero_static.py:
```diff
@@ -0,0 +1,65 @@
+import torch
```
If you don't mind, I'd like to bikeshed where exactly we should put this operator for a bit.

It doesn't seem right to put random export-oriented experimental operators in functorch. Functorch is about function transformations, which is why higher order operators should live there; nonzero doesn't really fit the bill here.

At the same time, there is no obvious preexisting spot to put it. We currently define custom operators in torch/_prims/__init__.py and torch/ao/quantization/fx/_decomposed.py.

I think there's a decent case to be made that nonzero_static makes sense as a prim. What do you think about having it live in `torch/_prims/__init__.py`?

guangy10 replied:
Yeah, I don't feel it belongs to functorch, but no idea where is the best place for it. Since you suggested `torch/_prims/__init__.py`, a few questions about it:
 1. Are prim ops "final" ops that can not be further decomposed? If yes, and we want to put `nonzero_static` here, it means `nonzero_static` is not decomposable?
 2. No idea which type of op `nonzeor_static` belongs to, I currently put it under "Tensor Creation Prims"

ezyang replied:
Yes, prim ops are final, not intended to be decomposed. It doesn't seem like you want to decompose nonzero_static anyway?

Tensor creation is not the right spot, but whatever, make a new section for it.

ezyang commented on test/dynamo/test_export.py:
```diff
@@ -2275,6 +2275,30 @@ def forward(self, *args):
         res = gm(input_tensor, input_tensor2)
         self.assertTrue(torch._dynamo.utils.same(ref, res))
 
+    @config.patch(dynamic_shapes=True)
```
I don't see why you need dynamic shapes = True; the whole point of static nonzero is that it works without dynamic shapes.

guangy10 replied:
Yeah, need clarification here. 

IIUC we will (1) set `dynamic_shapes=True` by default for export eventually (2) we want to make the program using `nonzero_static()` exportable. That's why it's set True here.

Secondly the program with `nonzero_static()` is exportable (no error) when `dynamic_shapes=False` **w/o adding the meta function for it**, so if it's expected to be used with `dynamic_shapes=False` why bother adding a meta function then?

May not direct relevant to this comment, but I want to ask if we would like to make the `size` param `int` or `SymInt`? Currently I assume we would always want to burn the `size` statically to the exported program so making it a `int`, but I'm not sure whether we have use cases would like to make it `SymInt`

guangy10 replied:
> I don't believe you lol. How did you repro this.

@ezyang It's reproducible for both c++ impl and py impl. Here is a patch (D44514255) for c++ impl (ignore the dtoh sync problem). As you can see I explicitly set dynamic_shapes=False for export and commented out the meta function in _meta_registrations.py, the test passes.

ezyang commented on functorch/experimental/_nonzero_static.py:
```diff
@@ -0,0 +1,65 @@
+    `tensor([[0, 0], [1, 0])` (truncated)
+    """
+    # Check if 'fill_value' is an integer type
+    assert isinstance(fill_value, int), "'fill_value' must be an integer"
```
Use check from prim_utils rather than assert for user checks

guangy10 replied:
Which prim_utils? I don't see any match when I searched "prim_utils". By looking at `torch/_prims/__init__.py`, it looks like many ops there are actually using `assert`.

ezyang commented on functorch/experimental/_nonzero_static.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def meta_nonzero_static(self: torch.Tensor, *, size: int, fill_value=-1):
+    return self.new_empty((size, self.dim()))
```
The result is always an int tensor, even if the input is a float tensor, so new_empty here is wrong

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -3043,6 +3043,81 @@ def f(pred, x):
         b = opt_fn(torch.tensor(True), torch.tensor([0.25, 0.25]))
         self.assertTrue(same(torch.sin(torch.tensor([0.25, 0.25])), b))
 
+    def test_nonzero_static(self):
```
Instead of doing these tests by hand, you should write an opinfo for nonzero_static (it will be similar to the opinfo for nonzero, see torch/testing/_internal/common_methods_invocations.py grep for nonzero), and then you will get a number of tests for free. It's a little non-standard how to specify non-aten ops, but if you move nonzero_static to prims namespace, then you can follow the prims examples here.

guangy10 replied:
Cool! Will replace the manual test cases with `opinfo`.

guangy10 replied:
Get back to this today. Will write an `opinfo` for `nonzero_static`. Two questions:
 1. How to run the specific test (where is the test file?) for my op or `nonzero()`? I want to test it locally.
 2. There are bunch of [skipped cases](https://github.com/pytorch/pytorch/blob/477f3f555f2a7953a1d2c56b732ce59aca46ae15/torch/testing/_internal/common_methods_invocations.py#L16754-L16767) for `nonzero()`, how do I know what to skip for `nonzero_static()`?

guangy10 replied:
FYI, will remove this test later once able to verify all test cases here are covered by `opinfo`

guangy10 replied:
Chatted with Ed. OpInfo is not a full replacement of manual unit tests. We still need manual tests to ensure the correctness of the op. So will keep the tests and and OpInfo

guangy10 commented on test/dynamo/test_export.py:
```diff
@@ -2275,6 +2275,30 @@ def forward(self, *args):
+            def forward(self, x):
+                return torch.ops.experimental.nonzero_static(x, size=self.static_size)
+
+        input_tensor = torch.tensor([6, 8])
```
Are you saying to test the case with `input_tensor = torch.empty(0)`? Can you elaborate?

ezyang replied:
torch.tensor creates a tensor with two elements [6, 8]. If you run nonzero on it, you will always fill in all the entries. It is good to also test what happens when you do have a zero tensor.

guangy10 commented on torch/_prims/__init__.py:
```diff
@@ -2673,6 +2674,84 @@ def _scalar_tensor_aten(
+    `tensor([[0, 0], [1, 0])` (truncated)
+"""
+
+# TODO: add layout and pin_memory support
```
Will remove

guangy10 commented on torch/_prims/__init__.py:
```diff
@@ -2673,6 +2674,84 @@ def _scalar_tensor_aten(
+    result[:copy_size, :] = dyn_result[:copy_size, :]
+    return result
+
+    # if isinstance(scalar, complex) and (
+    #     dtype is None or not utils.is_complex_dtype(dtype)
+    # ):
+    #     raise TypeError("Complex scalar requires complex tensor dtype.")
+    # # Note that Mypy thinks torch.scalar can't accept a complex scalar
+    # return torch.scalar_tensor(scalar, dtype=dtype, device=device)  # type: ignore[arg-type]
```
Will remove

guangy10 reviewed (commented):
Moved the op impl to prims. Addressed some of the comments. Need further clarification for the rest

zou3519 commented:
This is a drive-by comment, I don't have context over what is happening here.

We've been getting a lot of requests from users for a static nonzero (among other things): https://github.com/pytorch/pytorch/issues/62320, https://github.com/pytorch/functorch/issues/256. It looks like this PR adds a static nonzero for export-only use cases, but at some point we should actually expose this as a proper frontend operation because there are other use cases beyond export (e.g. vmap).

ezyang commented:
So what's going on is that guangy10 is just adding an experimental nonzero static op. But this is just to unblock export, he is not signed up to write the actual CUDA kernel that would be static. So it is still not usable for users...

guangy10 commented:
Update PR title to avoid confusion.

guangy10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6929,6 +6929,29 @@ def sample_inputs_nonzero(op_info, device, dtype, requires_grad, **kwargs):
+def sample_inputs_nonzero_static(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+
+    sizes = ((), (S,), (S, S), (S, S, S), (S, 1, S), (S, 0, S))
```
Similar issue as mentioned in the comment below when the rank is 0.
```
FAILED test_proxy_tensor.py::TestProxyTensorOpInfoCPU::test_make_fx_exhaustive__prims_nonzero_static_cpu_float32 - AssertionError: Rank must be positive but got 0
```

guangy10 reviewed (commented):
Using OpInfo for op tests

ngimel commented on torch/_prims/__init__.py:
```diff
@@ -2963,4 +2967,87 @@ def _fft_c2r_aten(
+        return torch.empty(0, dtype=torch.long)
+
+    shape: ShapeType = [size, a.dim()]
+    strides = utils.make_contiguous_strides_for(shape)
```
regular `nonzero` returns discontiguous tensor (transposed, with strides (1,n)), it would be good for static nonzero to follow the pattern

guangy10 replied:
Any utils function to use, otherwise I guess I will just make the stride (1,n) directly

ezyang replied:
there are a few ways to do it but you can either do `empty_permuted` or you can make a regular tensor and transpose it (slightly worse; empty_permuted is better.)

ngimel commented on torch/_prims/__init__.py:
```diff
@@ -2963,4 +2967,87 @@ def _fft_c2r_aten(
+
+    if size == 0:
+        # Return empty tensor
+        return torch.empty(0, dtype=torch.long)
```
should `size=0` return 1d or 2d tensor? Regular `nonzero` returns 2d tensor if there are no non-zero elements, with the size (0, ndim)

guangy10 replied:
I think the returned tensor size needs to match the specified `size` strictly.

ngimel replied:
`size` is an integer that indicates number of non-zero elements, so returned tensor size (2d) can't match `size`.

ezyang replied:
Yeah, it's an argument of uniformity. If you get (3, ndim) at size=3, (2, ndim) at size=2 and (1, ndim) at size=1, you should get (0, ndim) at size=0.

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6929,6 +6929,29 @@ def sample_inputs_nonzero(op_info, device, dtype, requires_grad, **kwargs):
+        mixed[mask_t] = 0
+        inputs.append(mixed)
+
+    nonzero_sizes = [0, 1, XS, S, M]
```
You can specify `error_inputs` function to test that some inputs raise an error.

guangy10 commented:
Chatted with @ezyang, we decided to go back to the c++ impl instead of continue the crappy py impl considering out variant will be needed for it and `ops._prims` currently doesn't support yet. 

The goal for this c++ impl is to primarily unblock export. The work to write a cuda kernel to avoid dtoh sync will be separated (tracked in #62320)

ngimel commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+  int64_t copy_len = std::min(size, num_nonzero);
+  // Copy the dynamic result to the fixed-size tensor
+  result.narrow(0, 0, copy_len).copy_(dyn_result.narrow(0, 0, copy_len));
+  if (ndim > copy_len) {
+    // Pad result with `fill_value`
+    result.narrow(0, copy_len, ndim - copy_len).fill_(fill_value);
```
```suggestion
  if (num_nonzero > copy_len) {
    // Pad result with `fill_value`
    result.narrow(0, copy_len, num_nonzero - copy_len).fill_(fill_value);
```

guangy10 replied:
Oh, yes! Stupid bug. I will fix and check my test why it's caught

guangy10 commented on test/dynamo/test_misc.py:
```diff
@@ -3055,6 +3055,76 @@ def f(pred, x):
         b = opt_fn(torch.tensor(True), torch.tensor([0.25, 0.25]))
         self.assertTrue(same(torch.sin(torch.tensor([0.25, 0.25])), b))
 
+    def test_nonzero_static(self):
```
Is the out variant accessible from users? I tried call `torch.nonzero_static_out()` in the test, it said this can not be found. I think out variant can be accessed as function. No?

guangy10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16766,6 +16789,22 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+               # Same set of skips as dynamic nonzero() above
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_out'),
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_out_warning'),
+               DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
```
I note that there are "Unexpected success" for this test case. After looking into it, it turns out that this test is marked expected failure for regular `nonzero()` due to a [bug](https://github.com/pytorch/pytorch/issues/67458) of passing as_tuple to it, which shouldn't affect nonzero_static(). So will remove this from the `skips` list

guangy10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16766,6 +16789,22 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_out_warning'),
+               DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
+               DecorateInfo(unittest.expectedFailure, 'TestNormalizeOperators', 'test_normalize_operator_exhaustive'),
+               DecorateInfo(unittest.expectedFailure, 'TestOperatorSignatures', 'test_get_torch_func_signature_exhaustive'),
```
Same. Will remove

guangy10 commented:
Checked each failed jobs but none is relevant to the code. Is it common that GitHub CI fails randomly, e.g. under same job, tests on one machine passed but failed on others and no option to re-run...

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+  // Sanity check on input
+  TORCH_CHECK(size >= 0, "'size' must be an non-negative integer");
+  int64_t ndim = self.dim();
+  TORCH_CHECK(ndim > 0, "Rank must be positive");
```
OK, you know what, I changed my mind, we should allow zero rank as it generalizes cleanly.

guangy10 replied:
What is the expected behavior of the zero rank input? Same as regular `nonzero()`?

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+      result.dim() == 2,
+      "nonzero_static: Expected out tensor to be a 2D tensor but got a ",
+      result.dim(),
+      "D tensor");
```
This is not entirely consistent with the prevailing convention for out= variants. In particular, when you pass a zero size out, PyTorch traditionally resizes it to the requested output size. Use `at::native::resize_output` to get the correct behavior in these cases.

guangy10 replied:
@ezyang  Just to check, do we want resize only when passing a zero size out, or whenever the size is doesn't exactly meet the expected output size, e.g. larger or smaller?

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+  // Allocate fixed-size out tensor
+  auto result = at::full(
+      {size, self.dim()},
+      fill_value,
```
You explicitly pad inside the out function, so it's not necessary to fill here. Use at::empty instead.

ezyang reviewed (commented):
You need to add docs for the function, conventionally they go in torch/_tensor_docs.py. We also should bikeshed whether or not we're giving this the new name nonzero_static, or making it an overload on the existing nonzero. (I'm not sure, checking if @albanD has an opinion. Note that jax makes it an overload)

ezyang commented:
Alban's vote is for making it an overload

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+  if (size > copy_len) {
+    // Pad result with `fill_value`
+    result.narrow(0, copy_len, size - copy_len).fill_(fill_value);
+  }
```
There are a bunch of ways we could make this marginally faster, but it probably doesn't make sense absent actually fixing this to not call nonzero_cpu lol

guangy10 reviewed (commented):
- Make zero rank input valid, return a tensor with size=(size, 0)
 - Resize out tensor if it doesn't match expected size
 - Remove duplicate padding logic
 - Add new op to `_tensor_docs.py`

guangy10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2243,6 +2243,74 @@ Tensor nonzero_cpu(const Tensor& self) {
+  at::assert_no_internal_overlap(result);
+  at::assert_no_overlap(result, self);
+
+  // Delegate call to regular nonzero to get a data-dependent output
```
Another pain with github code review. It's hard to tell which line(s) the comment is associated to, as a consequence sometimes it's difficult to understand the comment

ezyang replied:
Hmm this is not the line I put this comment on. It could be because of force pushes. You should use ghexport instead

guangy10 commented on test/dynamo/test_misc.py:
```diff
@@ -2936,6 +2936,105 @@ def f(pred, x):
         b = opt_fn(torch.tensor(True), torch.tensor([0.25, 0.25]))
         self.assertTrue(same(torch.sin(torch.tensor([0.25, 0.25])), b))
 
+    def test_nonzero_static(self):
```
It seems like I can not call `torch.nonzero_static_out()`. @ezyang Is the out variant hidden from users, i.e. not accessible directly? I'm asking because I'm about to test the resizing logic here but realized that out variant is not accessible.

ezyang replied:
You use the nonzero_static name and pass an out= kwarg

guangy10 commented:
> We also should bikeshed whether or not we're giving this the new name nonzero_static, or making it an overload on the existing nonzero. (I'm not sure, checking if @albanD has an opinion. Note that jax makes it an overload)

I initially added as a `nonzero.static`, but I think @suo prefer `torch.nonzero_static()`. @albanD your option? And your preference @ezyang?

suo commented:
My preference on the matter is weak. If jax just makes it an overload and that's what users are trained to expect, that's fine by me.

guangy10 commented:
Explored about making it an overload of `nonzero()` and think of the use cases of it:
 1. Python binding is skipped for `nonzero()` in https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_python_functions.py#L135. It looks like it's quite complicated to make it correct manually. Not sure it worth the time.
 2. The static variant will always assume `as_tuple=False`, so the output format of it is not 100% same as the regular `nonzero()`. Calling the static variant vs regular variant is not always exchangeable. Using one or the other should be explicit and not misleading. 
 3. Since we don't have cuda impl yet, we may want to limit the usage of this op only when it's necessary, e.g. to unblock export by replacing `nonzero()` with `nonzero_static()`. Another reason I think it's better to make the func name explicit rather than overload.
 
 So I would think we should better making it `nonzero_static()`, similar to `nonzero_numpy()`. cc: @ezyang @albanD

ezyang commented:
ok sgtm

guangy10 commented:
Currently merge is blocked due to the failed CI. The reason is because the `kwargs` are not supported in `test_overrides.py`. The issue is addressed in #98396. Will merge this PR after it.

guangy10 commented:
Now #98396 is merged and synced to fbcode. Can start merging this PR now.

guangy10 commented:
Test coverage for nonzero_static.out

guangy10 commented:
@pytorchbot merge

guangy10 commented:
The failed macos job is irrelevant

@pytorchbot merge -ic

guangy10 commented:
@pytorchbot merge -ic

guangy10 commented:
Separated internal changes out from this PR. Will try merge again later.

guangy10 commented:
Fix test_dtensor_ops.py::TestDTensorOpsCPU::test_dtensor_op_db_nonzero_cpu_float32

guangy10 committed:
Add nonzero_static() as an experimental op for PT2 export (#97417)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/97417

Add new op (`torch.nonzero_static`) to pytorch.

Test Plan:
## Unit Tests
```
buck test mode/dev-nosan //caffe2/test:test_dynamo -- 'caffe2/test:test_dynamo - test_export.py::ExportTests::test_export_with_nonzero_static' -- 'caffe2/test:test_dynamo - test_misc.py::MiscTests::test_nonzero_static'
```
https://www.internalfb.com/intern/testinfra/testrun/4222124811332637

## PT2 Export with `nonzero_static()`
Example of `GraphModule` in the exported graph
```
def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    nonzero_static_default = torch.ops.prims.nonzero_static.default(arg0, size = 4);  arg0 = None
    return pytree.tree_unflatten([nonzero_static_default], self._out_spec)
```

Reviewed By: ezyang

Differential Revision: D44324808

fbshipit-source-id: 8d8a76f345d1479edf6296387da692cdd49eb6e4

guangy10 commented:
Added `nonzero_static` to `UNIMPLEMENTED_XFAILLIST`.

@ezyang  The failed job is macos specific, so I'm not able to run the test `test_mps.py` from my devserver. What are the tips to debug such platform specific failure?

ezyang commented:
assuming your macbook supports mps, check out https://dev-discuss.pytorch.org/t/a-small-mps-debugging-story/769

guangy10 commented:
@pytorchbot merge -ic

guangy10 commented:
@pytorchbot merge -ic

netanel-haber commented:
@guangy10 Thank you for `fill_value` :heart_decoration:
